pkg/attribute/rootBag.go

```diff
@@ -0,0 +1,224 @@
+// Copyright 2016 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package attribute
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/golang/protobuf/ptypes"
+
+	mixerpb "istio.io/api/mixer/api/v1"
+)
+
+// The rootBag is always the base of a chain of bags. It's state is updated
+// only via an Attributes proto. This code is not protected by locks because
+// we know it is always mutated in a single-threaded context.
+type rootBag struct {
```

Comment:
This seems like a lot of code for a performance improvement that may not be significant or necessary. A null check is negligible compared to a map lookup, so I wouldn't think testing for ab.parent == nil would be noticeable.